### PR TITLE
#1(with refactoring and partly of tasks 2-4)

### PR DESCRIPTION
**Pull request by sinovdeath**
In feature/1 I refactored the architecture: a database and a service with validation were added. New fields "category"(mandatory), "discount"(optional) and "desctription"(optional) were added in the Product.
Also CreateRequest and CreateResponse were added. CreateRequest is responsible for method AddProductToDB in the Service and if request is valide in the CreateResponse, then Service create Product and transfer it to DB. If it invalide, then CreateResponse displays Errors.
Also the validation for fields "name", "price", "category", "discount" and "description" were implemented in the Validation class. Fields name, price and category has mandatory priority, discount and description has optional priority. "name", "price" and "category" checked for the empty field. "name" checked for the length violation from 3 to 32 symbols. "price" and "discount" checked for negative and higher than zero number, also "discount" checked for the range from 0 to 100 and if "price" lower than 20 euro, than discount can't be applied. "description" checked for the length violation from 8 to 60 symbols.
 Tests for Service, Validation and Database is covered 100% of code(method create in DB, and addToDB from service). In the future it is planned to add tests as methods are developed.

### CODE REVIEW

File: .idea/codeStyles/Project.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (javagururt) [inline .idea/codeStyles/Project.xml]: `.idea`  нужно удалить из гита. Потом добавить .gitignore

File: Test/com/javaguru/shoppinglist/Database/ProductRepositoryImplTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        milk = new Product("Milk", new BigDecimal("2.8"), ProductCategory.MILK);
+        milk.setProductDiscount(new BigDecimal("45.3"));
+        milk.setProductDescription("Good cow milk from Latvia");
+        meat = new Product("Meat", new BigDecimal("5.8"), ProductCategory.MEAT);
+        bread = new Product("Bread", new BigDecimal("1.5"), ProductCategory.BREAD);
+    }
+
+    @Test
+    public void testForCreateProductInDatabase() {
+        ProductRepositoryImpl db = new ProductRepositoryImpl();
```
Comment (javagururt) [inline Test/com/javaguru/shoppinglist/Database/ProductRepositoryImplTest.java]: создать базу  (цель тестирования) можно вне метода

File: Test/com/javaguru/shoppinglist/Service/ValidationTest.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+        String expectedError = "Name can't be empty";
+
+        List<ValidationErrors> list = validate.validateCreateRequest(milkWithoutName);
+
+        for (ValidationErrors validationErrors : list)
+        {
+            if (validationErrors.getResponse().equals(expectedError))
+            {
+                System.out.println("Test for validation of the create request with empty name has passed");
```
Comment (javagururt) [inline Test/com/javaguru/shoppinglist/Service/ValidationTest.java]: System.out.println это не проверка.

Comment (javagururt) [reply Test/com/javaguru/shoppinglist/Service/ValidationTest.java]: используй assert*

File: Test/com/javaguru/shoppinglist/Service/ValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+        List<ValidationErrors> list = validate.validateCreateRequest(milkWithoutPrice);
+
+        for (ValidationErrors validationErrors : list)
+        {
+            if (validationErrors.getResponse().equals(expectedError))
+            {
+                System.out.println("Test for validation of the create request with empty price has passed");
+            } else
+            {
```
Comment (javagururt) [inline Test/com/javaguru/shoppinglist/Service/ValidationTest.java]: форматирование поехало

File: src/com/javaguru/shoppinglist/Catalog/Product/Product.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
+package com.javaguru.shoppinglist.Catalog.Product;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+public class Product {
+    private static Long countID = 0L;
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/Product.java]: Неочевидная проперти

File: src/com/javaguru/shoppinglist/Catalog/Product/Product.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    private String productDescription = "";
+
+    public Product(String productName, BigDecimal productPrice, ProductCategory productCategory) {
+        this.productID = getCountID();
+        this.productName = productName;
+        this.productPrice = productPrice;
+        this.productCategory = productCategory;
+    }
+
+    private Long getCountID() {
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/Product.java]: геттер должен просто возвращать значение (не надо изменять)

File: src/com/javaguru/shoppinglist/Catalog/Product/ProductCategory.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public enum ProductCategory {
+    ALCOHOL, BREAD, FISH, FRUITS, MEAT, MILK, SOFT_DRINKS, SWEETS, VEGETABLES;
+
+    public static List<ProductCategory> getCategories() {
+        List<ProductCategory> categories = new ArrayList<>(Arrays.asList(ProductCategory.values()));
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/ProductCategory.java]: можно просто получать values и все. Нет смысла от метода особо

File: src/com/javaguru/shoppinglist/Catalog/Product/Request/CreateRequest.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    private String productCategory;
+
+    private BigDecimal productDiscount;
+    private String productDescription;
+
+    public CreateRequest() {
+        this.productName = null;
+        this.productPrice = null;
+        this.productCategory = null;
+        this.productDiscount = new BigDecimal("0.0");
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/Request/CreateRequest.java]: Есть константа `BigDecimal.ZERO`

File: src/com/javaguru/shoppinglist/Catalog/Product/Request/CreateRequest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+public class CreateRequest {
+    private String productName;
+    private BigDecimal productPrice;
+    private String productCategory;
+
+    private BigDecimal productDiscount;
+    private String productDescription;
+
+    public CreateRequest() {
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/Request/CreateRequest.java]: присваивать значения в конструкторе очень неочевидный подход.

File: src/com/javaguru/shoppinglist/Catalog/Product/Request/UpdateRequest.java [hunk 10]
```diff
@@ -0,0 +1,4 @@
+package com.javaguru.shoppinglist.Catalog.Product.Request;
+
+public class UpdateRequest {
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Catalog/Product/Request/UpdateRequest.java]: пустые классы лучше удалять

File: src/com/javaguru/shoppinglist/Service/Validation.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            return errorsList;
+        }
+        errorsList.addAll(lengthCheck(name));
+        return errorsList;
+    }
+
+    private List<ValidationErrors> lengthCheck(String name) {
+        List<ValidationErrors> errorsList = new ArrayList<>();
+
+        if (name != null
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Service/Validation.java]: лучше разделить проверки. Иначе условие выглядит сложно

File: src/com/javaguru/shoppinglist/Service/Validation.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            errorsList.add(ValidationErrors.NEGATIVE_OR_ZERO_PRICE);
+        }
+        return errorsList;
+    }
+
+    private List<ValidationErrors> validateToMoreThanTwentyPrice(BigDecimal price, BigDecimal discount) {
+        List<ValidationErrors> errorsList = new ArrayList<>();
+
+        if (price != null
+            && price.compareTo(BigDecimal.valueOf(20)) < 0
```
Comment (javagururt) [inline src/com/javaguru/shoppinglist/Service/Validation.java]: сложное условие

**Comment by sinovdeath** [comment]
> In feature/1 I refactored the architecture: a database and a service with validation were added. New fields "category"(mandatory), "discount"(optional) and "desctription"(optional) were added in the Product.
> Also CreateRequest and CreateResponse were added. CreateRequest is responsible for method AddProductToDB in the Service and if request is valide in the CreateResponse, then Service create Product and transfer it to DB. If it invalide, then CreateResponse displays Errors.
> Also the validation for fields "name", "price", "category", "discount" and "description" were implemented in the Validation class. Fields name, price and category has mandatory priority, discount and description has optional priority. "name", "price" and "category" checked for the empty field. "name" checked for the length violation from 3 to 32 symbols. "price" and "discount" checked for negative and higher than zero number, also "discount" checked for the range from 0 to 100 and if "price" lower than 20 euro, than discount can't be applied. "description" checked for the length violation from 8 to 60 symbols.
> Tests for Service, Validation and Database is covered 100% of code(method create in DB, and addToDB from service). In the future it is planned to add tests as methods are developed.

Merge feature/1.1 to feature/1

Validation for unique name.

**Pull request merged**